lib/spack/spack/schema/env.py

```diff
@@ -34,6 +35,8 @@
                 "include": {"type": "array", "default": [], "items": {"type": "string"}},
                 "specs": spec_list_schema,
                 "include_concrete": include_concrete,
+                "environment_variable_modifications": spack.schema.environment.definition,
```

Comment:
We're moving pretty much everything out of here and into its own config "section" to make it composable. If something only lives as an attribute here, in the environment file schema, it doesn't end up merging properly when we include it. Envs are going to be includable so we really want everything to be a proper config section. See these for context:

* `view.py`, which used to live here (#42250)
* Including envs (#46973)
* `include:` as config (#46792)

This should go in its own top-level config file.  I'd recommend calling it `environment:` for brevity, but maybe people feel like that is overloaded.  I think it will be clear when read.

I think the right organization is something like:
* rename current `environment.py` to `environment_modifications.py` (which is clearer... it's just the list of mods)
* move make a new `environment.py` that just defines:
    ```python
    {
        "environment": spack.schema.environment_modifications.definition,
    }
    ```
    like the section currently in `compilers.py`
* add an "environment" section to `SECTION_SCHEMAS` in `spack.config`, and to `spack.schema.merged.properties`.

Reply:
@tgamblin FWIW, I'm proposing we change this name to `shell_modifications`.  I feel that `environment` is already too overloaded when trying to explain spack `environment` vs the run environment, vs the build environment vs the dirty environment etc etc.

In terms of a schema I went back and forth on how to propose this to @fryeguy52  when starting.  The intent of this PR as I see it is to make modifications for this current `spack.yaml` driven environment.  In this sense it makes sense to me that they should be properties of the environment, but also composable if an env includes another env.  This puts them in the same vein as the specs config section IMO in that they could technically be a full schema, but they also shouldn't be used outside the environment.

I suppose that is debatable if we wanted to have shell mods that get applied when spack shell support is enabled or in spack subprocesses all the time.

What would you advise here? Still make this a full top level schema?

Reply:
> What would you advise here? Still make this a full top level schema?

Yes, so we don't have two ways to include things.  We're plannig to make `specs` into config, as well -- and it'll only really matter in an environment.

> FWIW, I'm proposing we change this name to `shell_modifications`. I feel that environment is already too overloaded when trying to explain spack environment vs the run environment, vs the build environment vs the dirty environment etc etc.

I get that it's overloaded but there are only really two meanings here:
* A spack environment (yeah maybe not the greatest name choice but we're kind of stuck with it now)
* A process's environment (build environment, run environment, etc. are all this)

I don't really like `shell_modifications` because this doesn't actually modify the shell. It modifies the activating process's _environment_... which is a pretty well understood concept.  What about `env_vars:` or `environment_vars:` or something like that?

Reply:
Thinking about this name more. I think what is confusing is we have different sets of modifications to a process environment. So the qualification is "the process environment with build modifications" or "the process environment with runtime modifications" which we just call the build or runtime environment.

Reply:
Ok per the discussion in an otherwise unrelated meeting today, we settled on calling the section `env_vars:`.  In the other places we have sections like this, we should rename those to `env_vars:` as well, but still support the `environment:` name for a while so we don't break people.

Reply:
I updated the name to `env_vars` and moved the schema to its own config file that can be included in a spack.yaml.